src/middlewared/middlewared/plugins/account_/2fa.py

```diff
@@ -37,10 +40,67 @@ def verify_twofactor_token(self, username, token):
         if not twofactor_config['enabled']:
             raise CallError('Please enable Two Factor Authentication first')
 
-        user = self.middleware.call_sync('user.query', [['username', '=', username]], {'get': True})
-        user_twofactor_config = self.middleware.call_sync('auth.twofactor.get_user_config', user['id'])
+        user = self.middleware.call_sync('user.translate_username', username)
+        if not user['twofactor_auth_configured']:
+            raise CallError('Two Factor Authentication is not configured for this user')
+
+        user_twofactor_config = self.middleware.call_sync(
+            'auth.twofactor.get_user_config', user['id' if user['local'] else 'sid'], user['local'],
+        )
         totp = pyotp.totp.TOTP(
             user_twofactor_config['secret'], interval=twofactor_config['interval'],
             digits=twofactor_config['otp_digits'],
         )
         return totp.verify(token or '', valid_window=twofactor_config['window'])
+
+    @private
+    async def translate_username(self, username):
+        """
+        Translates `username` to a user object.
+        """
+        try:
+            user = await self.middleware.call('user.get_user_obj', {'username': username, 'sid_info': True})
+        except KeyError:
+            raise CallError(f'User {username!r} does not exist', errno.ENOENT)
+
+        return await self.middleware.call(
+            'user.query', [['username', '=', user['pw_name']]], {
+                'get': True,
+                'extra': {'additional_information': ['DS']},
+            }
+        )
+
+    @accepts(Str('username'))
+    @returns(Ref('user_entry'))
+    async def renew_2fa_secret(self, username):
+        """
+        Renew `username` user's two-factor authentication secret.
+        """
+        user = await self.translate_username(username)
+        twofactor_auth = await self.middleware.call(
+            'auth.twofactor.get_user_config', user['id' if user['local'] else 'sid'], user['local']
+        )
+
+        # Add some sanity checks here
+        if user['local'] and not twofactor_auth['exists']:
+            raise CallError(f'Unable to locate two factor authentication configuration for {username!r} user')
+
+        secret = await self.middleware.call('auth.twofactor.generate_base32_secret')
```

Comment:
We should only generate the `secret` if the `if two factor_auth['exists']:` branch is true

Reply:
@yocalebo for AD users we do not add their record unless they actually configure 2fa which is why the check is for local users only as all local users are mapped to the 2fa table because we have them in the database already.
I'll add some comments to outline this behaviour, pretty sure i'll forget this myself as well after a while